pkg/skaffold/build/parallel.go

```diff
@@ -21,96 +21,103 @@ import (
 	"context"
 	"fmt"
 	"io"
+	"sync"
 
 	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/build/tag"
 	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/color"
 	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/event"
 	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/schema/latest"
-	"github.com/pkg/errors"
 )
 
 const bufferedLinesPerArtifact = 10000
 
 type artifactBuilder func(ctx context.Context, out io.Writer, artifact *latest.Artifact, tag string) (string, error)
 
 // InParallel builds a list of artifacts in parallel but prints the logs in sequential order.
-func InParallel(ctx context.Context, out io.Writer, tags tag.ImageTags, artifacts []*latest.Artifact, buildArtifact artifactBuilder) ([]Artifact, error) {
+func InParallel(ctx context.Context, out io.Writer, tags tag.ImageTags, artifacts []*latest.Artifact, buildArtifact artifactBuilder) ([]chan Result, error) {
 	if len(artifacts) == 1 {
 		return InSequence(ctx, out, tags, artifacts, buildArtifact)
 	}
 
+	var resultChans []chan Result
+
 	ctx, cancel := context.WithCancel(ctx)
 	defer cancel()
 
-	n := len(artifacts)
-	finalTags := make([]string, n)
-	errs := make([]error, n)
-	outputs := make([]chan []byte, n)
+	// for collecting all output and printing in order later on
+	outputs := make([]chan []byte, len(artifacts))
+
+	for i, a := range artifacts {
+		resChan := make(chan Result, 1)
```

Comment:
Why not make a channel of results at L42 and avoid `append(resultChans, resChan)`?

Reply:
just creating a slice of channels isn't enough because each channel will still be nil. they need to be created individually.

I could do
```golang
resultChans := make([]chan Result, len(artifacts))
for i := 0; i < len(artifacts); i++ {
  resultChans[i] = make(chan Result, 1)
}
```
but I don't think this is much better.

Reply:
the reason why this is better is, append creates a new memory allocation and is little bit inefficient especially in this case when we know the size.

Reply:
updated